src/keys/bip39.rs

```diff
@@ -5,34 +5,216 @@
 // https://doc.rust-lang.org/std/primitive.str.html
 // "String slices are always valid UTF-8."
-type Mnemonic = str;
-type Passphrase = str;
-type Seed = [u8; 64];
 
 use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use core::convert::TryFrom;
+use core::ops::Deref;
 
-use unicode_normalization::UnicodeNormalization;
+use unicode_normalization::{is_nfkd, UnicodeNormalization};
+use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
 
-pub fn mnemonic_to_seed(m: &Mnemonic, p: &Passphrase, s: &mut Seed) {
-    let m = m.chars().nfkd().collect::<String>();
+/// BIP39 coded error.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    /// Mnemonic entropy amount is invalid (should be 128 or 160 or 192 or 224 or 256 bits)
+    InvalidEntropyCount(usize),
+    /// Mnemonic contains a word not present in word list
+    NoSuchWord(String),
+    /// Mnemonic corrupted, checksum mismatch
+    ChecksumMismatch,
+    /// Mnemonic is not in NFKD form
+    UnnormalizedMnemonic,
+    /// Passphrase is not in NFKD form
+    UnnormalizedPassphrase,
+    /// Word list contains unnormalized word or word with a separator
+    BadWordlistWord(String),
+    /// Word list contains duplicate words
+    UnsortedWordlist,
+    /// Separator is not in NFKD form
+    BadSeparator,
+}
+
+/// Reference to a normalized (unicode NFKD) mnemonic.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct MnemonicRef<'a>(&'a str);
+
+impl<'a> Deref for MnemonicRef<'a> {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.0
+    }
+}
+
+impl<'a> TryFrom<&'a str> for MnemonicRef<'a> {
+    type Error = Error;
+    fn try_from(mnemonic_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(mnemonic_str) {
+            Ok(MnemonicRef(mnemonic_str))
+        } else {
+            Err(Error::UnnormalizedMnemonic)
+        }
+    }
+}
+
+impl<'a> PartialEq<str> for MnemonicRef<'a> {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+/// Owned normalized (unicode NFKD) mnemonic.
+///
+/// Mnemonic is the encoding of secret entropy using words from a given word list.
+/// Mnemonic is used to derive a seed which serves as a master key.
+/// If mnemonic is leaked then the seed is compromised (unless a strong passphrase is used).
+/// Mnemonic should be kept secret on analog media.
+/// Mnemonic should be verified against a given word list before deriving a seed from it.
+#[derive(Clone, Zeroize, ZeroizeOnDrop)]
+pub struct Mnemonic(String);
 
-    let mut salt = String::with_capacity("mnemonic".len() + p.len());
-    salt.push_str("mnemonic");
-    salt.push_str(p);
-    let salt = salt.nfkd().collect::<String>();
+/// Normalize the input string and use it as mnemonic.
+/// The resulting mnemonic should be verified against a given word list before deriving a seed from it.
+impl From<String> for Mnemonic {
+    fn from(mut unnormalized_mnemonic: String) -> Self {
+        let mnemonic = Self(unnormalized_mnemonic.chars().nfkd().collect());
```

Comment:
I am actually astonished that this normalization does work for any `String` (i.e. it is infallible). What would a completely invalid string like 
"    abcdef   " (with leading and trailing whitespace) result in for example?

Reply:
Why is not something necessary like this?:
```rust
impl TryFrom<String> for Mnemonic {
    type Error = Error;

    fn try_from(mut value: String) -> Result<Self, Self::Error> {
        // trim because empty spaces could create a different seed https://github.com/iotaledger/crypto.rs/issues/125
        let trimmed = value.trim();
        // first we check if the mnemonic is valid to give meaningful errors
        if let Err(err) = crypto::keys::bip39::wordlist::verify(trimmed, &crypto::keys::bip39::wordlist::ENGLISH) {
            value.zeroize();
            Err(crate::client::Error::InvalidMnemonic(format!("{err:?}")))
        } else {
            let mnemonic = trimmed.to_string();
            value.zeroize();
            Ok(Self(mnemonic))
        }
    }
}
```

Reply:
oh ... I missed this line in the docs "Mnemonic should be verified against a given word list before deriving a seed from it." So we allow invalid `Mnemonic`s currently, right?

Reply:
This is unicode NFKD normalization. It is needed to have a unique representation of the same string written in different unicode characters. This `from` conversion doesn't check for spaces and words. This mnemonic should be verified against the intended word list. I've added a doc comment regarding this.

Reply:
Yeah, that's what I missed and saw eventually. So you think it's better to lazily verify the mnemonic (when actually needed for seed derivation) rather than eagerly in the `From` or then probably `TryFrom` impl?

Reply:
Not really sure. You can derive seed from any string/mnemonic even if it's not a valid according to some standard word lists. It would probably be better to validate mnemonic, but at this point (in `from` and `try_from` we don't have any word list to validate against). I'm fine with letting users use mnemonics without validation if that's what they want.

Reply:
can be closed